ast/ast.hpp

```diff
@@ -173,18 +179,45 @@ struct IfStatement : Statement {
     sPtr<Body> elseBody;
     bool operator==(const IfStatement& other) const {
         return Node::operator==(other) && condition == other.condition &&
-               ifBody == other.elseBody && ifBody == other.elseBody;
+               ifBody == other.ifBody && elseBody == other.elseBody;
     }
     void accept(Visitor& v) override { v.visit(this); }
 };
+
 struct Expression : Node {
+    sPtr<Expression> operand1;
+    sPtr<Expression> operand2;
+    lexer::Token operation;
+
+    bool operator==(const Expression& other) const {
+        return Node::operator==(other) && operand1 == other.operand1 &&
+               operand2 == other.operand2 && operation == other.operation;
+    }
     void accept(Visitor& v) override { v.visit(this); }
 };
-struct UnaryExpression : Expression {
+struct Primitive : Expression {
+    lexer::Token value;
+    bool operator==(const Primitive& other) const {
+        return Node::operator==(other) && value == other.value;
+    }
     void accept(Visitor& v) override { v.visit(this); }
 };
-struct BinaryExpression : Expression {
+struct ModifiablePrimary : Expression {
+    std::vector<sPtr<Expression>> args;
+
+    bool operator==(const ModifiablePrimary& other) const {
+        return Node::operator==(other) && args == other.args;
+    }
     void accept(Visitor& v) override { v.visit(this); }
 };
+struct RoutineCall : Expression {
+    lexer::Token routine;
```

Comment:
Better make it a pointer to the actual routine definition node

Reply:
Routine call can be used as a part of expression call, that is the reason why `RoutineCall` inherits from `Expression`

Reply:
This field of ast can be added during semantic analysis.

Reply:
Sure, but I was referring to `lexer::Token routine`. I meant that it is better to make `routine` a pointer to the actual routine declaration instead of just holding the name, like `sPtr<Routine> routine;`